src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -239,15 +242,22 @@ trait FastSync {
           blacklist(peer.id, blacklistDuration, s"responded with block bodies not matching block headers, blacklisting for $blacklistDuration")
           self ! FastSync.EnqueueBlockBodies(requestedHashes)
         case DbError =>
-          blockBodiesQueue = Seq.empty
-          receiptsQueue = Seq.empty
-          //todo adjust the formula to minimize redownloaded block headers
-          bestBlockHeaderNumber = bestBlockHeaderNumber - 2 * blockHeadersPerRequest
-          log.debug("missing block header for known hash")
-          self ! ProcessSyncing
+          redownloadBlockchain()
       }
     }
 
+    /**
+      * Restarts download from a few blocks behind the current best block header, as an unexpected DB error happened
+      */
+    private def redownloadBlockchain(): Unit = {
+      blockBodiesQueue = Seq.empty
+      receiptsQueue = Seq.empty
+      //todo adjust the formula to minimize redownloaded block headers
+      bestBlockHeaderNumber = bestBlockHeaderNumber - 2 * blockHeadersPerRequest
```

Comment:
You are right, nice catch! :+1:
The plan for the `TODO` is to investigate whether going back `2 * blockHeadersPerRequest` block headers is too much and maybe we could minimize it. However this DB problem should be really infrequent if there are no bugs, so we maybe could leave it as it is.